src/os/bluestore/BlueStore.cc

```diff
@@ -8108,18 +8118,22 @@ void BlueStore::_txc_finish(TransContext *txc)
   }
 }
 
-void BlueStore::_txc_release_alloc(TransContext *txc)
+void BlueStore::_txc_release_alloc_locked(TransContext *txc)
 {
-  // update allocator with full released set
+  // it's expected we're called with lazy_release_lock already taken!
   if (!cct->_conf->bluestore_debug_no_reuse_blocks) {
     dout(10) << __func__ << " " << txc << " " << txc->released << dendl;
-    for (interval_set<uint64_t>::iterator p = txc->released.begin();
-	 p != txc->released.end();
-	 ++p) {
-      alloc->release(p.get_start(), p.get_len());
-    }
+    // interval_set seems to be too costly for inserting things in
+    // bstore_kv_final. We could serialize in simpler format and perform
+    // the merge separately, maybe even in a dedicated thread.
+    lazy_release_extents.insert(txc->released);
   }
 
+  if (lazy_release_extents.num_intervals() >= 256) {
```

Comment:
Really need to find something better.

Reply:
instead of dong it here, only release if there is an atomic<bool> aggressive_release set.